active/0012-cluster-call.md

```diff
@@ -0,0 +1,201 @@
+               retry_catch_up_later(Reason)
+       end.
+   
+   catch_up(Node, ToTnxId) ->
+       case transaction(fun() -> do_catch_up(Node, ToTnxId) end) of
+           {atomic, caught_up} -> ok;
+           {atomic, still_lagging} -> catch_up(Node, ToTnxId);
+           {aborted, Reason} -> retry_catch_up_later(Reason)
+       end.
+   
+   do_catch_up(Node, ToTnxId) ->
+       case mnesia:wread({?CLUSTER_CURSOR, Node}) of
+           [] -> caught_up;
+           [#cluster_call_cursor{tnx_id = DoneTnxId}] when ToTnxId =< DoneTnxId -> caught_up;
+           [Rec = #cluster_call_cursor{tnx_id = DoneTnxId}] ->
+               CurTnxId = DoneTnxId + 1,
+               mnesia:write(Rec#cluster_call_cursor{tnx_id = CurTnxId}),
+               [#cluster_call_mfa{mfa = MFA}] = mnesia:read(?CLUSTER_MFA, CurTnxId),
+               apply_mfa(Node, MFA)
+       end.
+   ```
+
+3. If a new update operation is added, the `handler` will receive a write event of the `cluster_call_mfa` table.
+   When  `EventTxnId = DoneTnxId+1`, i.e. we have finished  transaction 4 and received the notification of transaction 5, we will execute transaction 5 directly, if we receive transaction 6, we will do nothing and let `catch_up` catch up from transaction 5.
+
+   ```erlang
+   handle_info({write, MFARec, _ActivityId}, State) ->
+       Node = node(),
+       case transaction(fun() -> handle_table_write_event(Node, MFARec) end) of
+           {atomic, catch_up} -> {noreply, State, {continue, catch_up}};
+           {abort, _Reason} -> {noreply, State, {continue, catch_up}};
+           {atomic, ok} -> {noreply, State, ?TIMEOUT}
+       end.
+   
+   handle_table_write_event(Node, #cluster_call_mfa{tnx_id = EventTnxId, mfa = MFA}) ->
+       DoneTnxId =
+           case mnesia:wread({?CLUSTER_CURSOR, Node}) of
+               [] -> EventTnxId - 1;
+               [#cluster_call_cursor{tnx_id = TnxId}] -> TnxId
+           end,
+       case EventTnxId =:= DoneTnxId + 1 of
+           false -> catch_up; %$% catch up latest id in catch_up/0
+           true ->
+               mnesia:write(#cluster_call_cursor{tnx_id = EventTnxId, node = Node}),
+               apply_mfa(Node, MFA)
+       end.
+   ```
+
+   
+
+4. The first transaction must be executed in the `emqx_cluster_trans` process. if this transaction succeeds, the call returns success directly, if the transaction fails, the call aborts with failure.
+
+   ```erlang
+   commit(Node, MFA) ->
+       case catch_up() of
+           retry -> {error, "catch up failed"};
+           ok ->
+               Trans = fun() -> apply_to_commit(Node, MFA) end,
+               transaction(Trans)
+       end.
+   apply_to_commit(Node, MFA) ->
+       mnesia:write_lock_table(?CLUSTER_MFA),
+       #cluster_call_cursor{tnx_id = LastDoneId} = mnesia:read(?CLUSTER_CURSOR, Node),
+       #cluster_call_mfa{tnx_id = LatestId} = mnesia:last(?CLUSTER_MFA),
+       ok = do_catch_up_in_one_trans(LatestId, LastDoneId, Node),
```

Comment:
we need some comment here to explain why there should be a catch up again after line 126.